Source/Classes/PINRemoteImageManager.m

```diff
@@ -1286,28 +1219,17 @@ - (NSUUID *)downloadImageWithURLs:(NSArray <NSURL *> *)urls
             }
         }];
         
-        float currentBytesPerSecond = [strongSelf currentBytesPerSecond];
-        [strongSelf lock];
-            float highQualityQPSThreshold = [strongSelf highQualityBPSThreshold];
-            float lowQualityQPSThreshold = [strongSelf lowQualityBPSThreshold];
-            BOOL shouldUpgradeLowQualityImages = [strongSelf shouldUpgradeLowQualityImages];
-        [strongSelf unlock];
-        
-        NSUInteger desiredImageURLIdx;
-        if (currentBytesPerSecond == -1 || currentBytesPerSecond >= highQualityQPSThreshold) {
-            desiredImageURLIdx = urls.count - 1;
-        } else if (currentBytesPerSecond <= lowQualityQPSThreshold) {
-            desiredImageURLIdx = 0;
-        } else if (urls.count == 2) {
-            desiredImageURLIdx = roundf((currentBytesPerSecond - lowQualityQPSThreshold) / ((highQualityQPSThreshold - lowQualityQPSThreshold) / (float)(urls.count - 1)));
-        } else {
-            desiredImageURLIdx = ceilf((currentBytesPerSecond - lowQualityQPSThreshold) / ((highQualityQPSThreshold - lowQualityQPSThreshold) / (float)(urls.count - 2)));
-        }
+        NSUInteger desiredImageURLIdx = [strongSelf appropriateImageIdxForURLsGivenHistoricalNetworkConditions:urls];
         
         NSUInteger downloadIdx;
         //if the highest quality already downloaded is less than what currentBPS would dictate and shouldUpgrade is
         //set, download the new higher quality image. If no image has been cached, download the image dictated by
         //current bps
+        
+        [strongSelf lock];
+            BOOL shouldUpgradeLowQualityImages = [strongSelf shouldUpgradeLowQualityImages];
+        [strongSelf unlock];
```

Comment:
This may well be a micro-optimization, but to avoid grabbing the lock multiple times, how about taking this chance to get high and low quality thresholds and pass them to `-appropriateImageIdxForURLsGivenHistoricalNetworkConditions`? Then that method can be static because it no longer needs access to `self`.